src/sources/aws_s3/mod.rs

```diff
@@ -0,0 +1,476 @@
+        let s3_client = S3Client::new_with(
+            client.clone(),
+            Arc::<rusoto::AwsCredentialsProvider>::clone(&creds),
+            region.clone(),
+        );
+
+        match self.sqs {
+            Some(ref sqs) => {
+                let sqs_client = SqsClient::new_with(
+                    client.clone(),
+                    Arc::<rusoto::AwsCredentialsProvider>::clone(&creds),
+                    region.clone(),
+                );
+
+                sqs::Ingestor::new(
+                    region.clone(),
+                    sqs_client,
+                    s3_client,
+                    sqs.clone(),
+                    self.compression,
+                    multiline,
+                )
+                .await
+                .context(Initialize {})
+            }
+            None => Err(CreateSqsIngestorError::ConfigMissing {}),
+        }
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum CreateSqsIngestorError {
+    #[snafu(display("Unable to initialize: {}", source))]
+    Initialize { source: sqs::IngestorNewError },
+    #[snafu(display("Unable to create AWS client: {}", source))]
+    Client { source: crate::Error },
+    #[snafu(display("Unable to create AWS credentials provider: {}", source))]
+    Credentials { source: crate::Error },
+    #[snafu(display("Configuration for `sqs` required when strategy=sqs"))]
+    ConfigMissing,
+    #[snafu(display("Could not parse region configuration: {}", source))]
+    RegionParse { source: rusoto::region::ParseError },
+}
+
+fn s3_object_decoder(
+    compression: Compression,
+    key: &str,
+    content_encoding: Option<&str>,
+    content_type: Option<&str>,
+    body: rusoto_s3::StreamingBody,
+) -> Box<dyn tokio::io::AsyncRead + Send + Unpin> {
+    use async_compression::tokio_02::bufread;
+
+    let r = tokio::io::BufReader::new(body.into_async_read());
+
+    let compression = match compression {
+        Auto => {
+            determine_compression(content_encoding, content_type, key).unwrap_or(Compression::None)
+        }
+        _ => compression,
+    };
+
+    use Compression::*;
+    match compression {
+        Auto => unreachable!(), // is mapped above
+        None => Box::new(r),
+        Gzip => Box::new(bufread::GzipDecoder::new(r)),
+        Zstd => Box::new(bufread::ZstdDecoder::new(r)),
+    }
+}
+
+/// try to determine the compression given the:
+/// * content-encoding
+/// * content-type
+/// * key name (for file extension)
+///
+/// It will use this information in this order
+fn determine_compression(
+    content_encoding: Option<&str>,
+    content_type: Option<&str>,
+    key: &str,
+) -> Option<Compression> {
+    content_encoding
+        .and_then(content_encoding_to_compression)
+        .or_else(|| content_type.and_then(content_type_to_compression))
+        .or_else(|| object_key_to_compression(key))
+}
+
+fn content_encoding_to_compression(content_encoding: &str) -> Option<Compression> {
+    use Compression::*;
```

Comment:
Just a minor nit, but for just two cases, and the fact that it requires the non-standard way of writing `Option::None`, I would forgo this line, and just write `Compression::Gzip` and `Compression::Zstd`.